server/server.go

```diff
@@ -6,53 +6,299 @@ package server
 
 import (
 	"context"
+	"errors"
 	"fmt"
-
-	pbts "google.golang.org/protobuf/types/known/timestamppb"
+	"io"
+	"sync"
+	"sync/atomic"
+	"time"
 
 	"github.com/elastic/elastic-agent-libs/logp"
 	pb "github.com/elastic/elastic-agent-shipper-client/pkg/proto"
 	"github.com/elastic/elastic-agent-shipper-client/pkg/proto/messages"
 	"github.com/elastic/elastic-agent-shipper/queue"
+
+	"github.com/gofrs/uuid"
 )
 
+// ShipperServer contains all the gRPC operations for the shipper endpoints.
+type ShipperServer interface {
+	pb.ProducerServer
+	io.Closer
+}
+
 type shipperServer struct {
 	logger *logp.Logger
+	queue  *queue.Queue
+	cfg    ShipperServerConfig
+
+	uuid           string
+	acceptedIndex  int64
+	persistedIndex int64
 
-	queue *queue.Queue
+	polling       polling
+	notifications notifications
+
+	close *sync.Once
 
 	pb.UnimplementedProducerServer
 }
 
-// PublishEvents is the server implementation of the gRPC PublishEvents call
-func (serv shipperServer) PublishEvents(_ context.Context, req *messages.PublishRequest) (*messages.PublishReply, error) {
-	results := []*messages.EventResult{}
-	for _, evt := range req.Events {
-		serv.logger.Infof("Got event %s: %#v", evt.EventId, evt.Fields.Data)
-		err := serv.queue.Publish(evt)
-		if err != nil {
-			// If we couldn't accept any events, return the error directly. Otherwise,
-			// just return success on however many events we were able to handle.
-			if len(results) == 0 {
-				return nil, err
+type polling struct {
+	ctx     context.Context
+	stop    func()
+	stopped chan struct{}
+}
+
+// NewShipperServer creates a new server instance for handling gRPC endpoints.
+func NewShipperServer(q *queue.Queue, cfg ShipperServerConfig) (ShipperServer, error) {
+	if q == nil {
+		return nil, errors.New("queue cannot be nil")
+	}
+
+	err := cfg.Validate()
+	if err != nil {
+		return nil, fmt.Errorf("invalid configuration: %w", err)
+	}
+
+	id, err := uuid.NewV4()
+	if err != nil {
+		return nil, err
+	}
+
+	s := shipperServer{
+		uuid:   id.String(),
+		logger: logp.NewLogger("shipper-server"),
+		queue:  q,
+		cfg:    cfg,
+		polling: polling{
+			stopped: make(chan struct{}),
+		},
+		notifications: notifications{
+			subscribers: make(map[uuid.UUID]chan change),
+			mutex:       &sync.Mutex{},
+		},
+		close: &sync.Once{},
+	}
+
+	s.polling.ctx, s.polling.stop = context.WithCancel(context.Background())
+	s.startPolling()
+
+	return &s, nil
+}
+
+// GetAcceptedIndex atomically reads the accepted index
+func (serv *shipperServer) GetAcceptedIndex() int64 {
+	return atomic.LoadInt64(&serv.acceptedIndex)
+}
+
+// GetPersistedIndex atomically reads the persisted index
+func (serv *shipperServer) GetPersistedIndex() int64 {
+	return atomic.LoadInt64(&serv.persistedIndex)
+}
+
+// PublishEvents is the server implementation of the gRPC PublishEvents call.
+func (serv *shipperServer) PublishEvents(_ context.Context, req *messages.PublishRequest) (*messages.PublishReply, error) {
```

Comment:
I don't think we can enforce a timeout here. We range events one by one and send them to the queue. If we suddenly time out this would mean data duplication on the next try, would it not?

Also, why the server should enforce a timeout? Should not it be the client setting the timeout on its end?

Reply:
Any timeout set by a gRPC client will propagated here in the context of the RPC handler as a deadline. We could look at the context and return an error if the client hasn't set a deadline.

I think in the context where the shipper runs as a process on the same host as the client this probably low value.